app/controllers/concerns/application_shared.rb

```diff
@@ -21,4 +21,82 @@ def set_timezone
+      determined_taxonomy = nil
+
+      if api_request? # API request
+        if user.admin? && !params.has_key?("#{taxonomy}_id") # admin always uses any context, they can limit the scope with explicit parameters such as organization_id(s)
+          determined_taxonomy = nil
+        elsif params.has_key?("#{taxonomy}_id") # admin and non-admin who specified explicit context
+          determined_taxonomy = available.where(:id => params["#{taxonomy}_id"]).first
+
+          # in case admin asked for taxonomy that does not exist or is not accessible, we reply with 404
+          if determined_taxonomy.nil?
+            not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
+            return false
+          end
+        elsif request.session["#{taxonomy}_id"].present? # non-admin who didn't not specify explicit context
+          determined_taxonomy = find_session_taxonomy(taxonomy, user)
+        else # non-admin user who didn't specify explicit context and does not have anything in session
+          determined_taxonomy = find_default_taxonomy(user, taxonomy)
+        end
+      else # UI request
+        if request.session["#{taxonomy}_id"].present?
+          determined_taxonomy = find_session_taxonomy(taxonomy, user)
+        elsif !user.admin? && available.count == 1
+          determined_taxonomy = available.first
+        end
+      end
+
+      store_taxonomy(taxonomy, determined_taxonomy) if determined_taxonomy.present?
+    end
+  end
+
+  def store_taxonomy(taxonomy, determined_taxonomy)
+    taxonomy.classify.constantize.send 'current=', determined_taxonomy
+    request.session["#{taxonomy}_id"] = determined_taxonomy.id
+  end
+
+  def store_default_taxonomies(user)
+    ['location', 'organization'].each do |taxonomy|
+      default_taxonomy = find_default_taxonomy(user, taxonomy)
+      store_taxonomy(taxonomy, default_taxonomy) if default_taxonomy.present?
+    end
+  end
+
+  # we want to be explicit to keep this readable
+  # rubocop:disable Style/EmptyElse
+  def find_default_taxonomy(user, taxonomy)
+    default_taxonomy = user.send "default_#{taxonomy}"
+    available = user.send("my_#{taxonomy.pluralize}")
+
+    if default_taxonomy.present?
+      default_taxonomy
+    elsif available.count == 1
+      available.first
+    else
+      # no default taxonomy and user is either admin or user with more taxonomies, nil represents "Any Context"
+      nil
+    end
+  end
+
+  def find_session_taxonomy(taxonomy, user)
+    available = user.send("my_#{taxonomy.pluralize}")
+    determined_taxonomy = available.where(:id => request.session["#{taxonomy}_id"]).first
+    # warn user if taxonomy stored in session does not exist and delete it from session (probably taxonomy has been deleted meanwhile)
+    if determined_taxonomy.nil?
+      warning _("#{taxonomy.capitalize} you had selected as your context has been deleted") unless api_request?
```

Comment:
Translating interpolated string, please use `format` instead

Reply:
good catch, fixed and tested the warning prints correctly, I also added integration tests for this path which revealed one more issue, keeping it in separate commit so you can see better